### ISSUE TITLE
Pyright warnings in strict mode: type is partially unknown for op.create_index and sa.Column (when sa.Numeric type is used)

### ISSUE THREAD

**Issue by AlexanderPodorov** [sqlalchemy/alembic#1190]
Two typing warnings in VSCode in strict mode (type is partially unknown):

Example of `op.create_index`:
`op.create_index(op.f('ix_token_expired_at'), 'token', ['expired_at'], unique=False)`
warning: Type of "create_index" is partially unknown
reason (most likely): `columns` parameter has type `Sequence[str | TextClause | Function[Unknown]]`. `Function` generic is missing type parameter.

Example of `sa.Column` with `sa.Numeric` (all other types seem to work well):
`sa.Column('tax_percentage', sa.Numeric(precision=8, scale=2), nullable=False),`
warning: Argument type is partially unknown
reason (most likely): `Numeric` class has following signature:
```python
_N = TypeVar("_N", bound=Union[decimal.Decimal, float])

class Numeric(HasExpressionLookup, TypeEngine[_N]):
    ...
```
I believe `_N` should be defined as `_N = TypeVar("_N", decimal.Decimal, float)`, so no `bound=` and `Union`.

**Expected behavior**
No warnings produced

**Versions.**
 - OS: MacOS
 - Python: 3.11.2
 - Alembic: 1.9.4
 - SQLAlchemy: 2.0.4
 - Database: Postgres 14
 - DBAPI: asyncpg

Thanks a lot for `sqlalchemy` and `alembic`!

**Comment by zzzeek** [comment]
we can accept PRs for this if you have resources

**Comment by AlexanderPodorov** [comment]
Unfortunately I don't, sorry!

**Comment by CaselIT** [comment]
I'm not sure if changing the typevar actually changes anything here, since typevar can't have default type as of yet. The issue seems to be a missing type definition.

I wonder adding inits overload like in Float would help here https://github.com/sqlalchemy/sqlalchemy/blob/da70478eb2eafe9c76b836217371e029c3c820e3/lib/sqlalchemy/sql/sqltypes.py#L610-L625

**Comment by AlexanderPodorov** [comment]
Maybe, it would help... I'm not sure.
I just wanted to mention that if a `TypeVar` can have multiple types, they should be enumerated instead of `Union`:
```python
_N = TypeVar("_N", decimal.Decimal, float)
```
https://docs.python.org/3/library/typing.html#typing.TypeVar
Whatever way that works, should be good.

**Comment by CaselIT** [comment]
that's not the same as bound though. In this use case it should be similar, but in general the two are not the same

**Comment by AlexanderPodorov** [comment]
Yes, I'm wondering is it correct use case to use `Union` with `bound=`. How do we define "any subtype" of `Union` of two types.

**Comment by CaselIT** [comment]
well, `class Foo(float): pass` is a subclass of the union

**Comment by CaselIT** [comment]
this is actually a sqlalchemy thing. moving there
